server_development/topics/auth-spi.adoc

```diff
@@ -108,199 +111,547 @@ Let's walk through the steps from when a client first redirects to keycloak to a
+  sub-flow, the user will never see the OTP login page, unless the Conditional OTP subflow is set to Required.
 . After the flow is complete, the authentication processor creates a UserSessionModel and associates it with the AuthenticationSessionModel.
-  It then checks to see if the user is required to complete any required actions before logging in. 
+  It then checks to see if the user is required to complete any required actions before logging in.
 . First, each required action's evaluateTriggers() method is called.
   This allows the required action provider to figure out if there is some state that might trigger the action to be fired.
-  For example, if your realm has a password expiration policy, it might be triggered by this method. 
+  For example, if your realm has a password expiration policy, it might be triggered by this method.
 . Each required action associated with the user that has its requiredActionChallenge() method called.
   Here the provider sets up an HTTP response which renders the page for the required action.
-  This is done by setting a challenge status. 
-. If the required action is ultimately successful, then the required action is removed from the user's require actions list. 
-. After all required actions have been resolved, the user is finally logged in.         
+  This is done by setting a challenge status.
+. If the required action is ultimately successful, then the required action is removed from the user's require actions list.
+. After all required actions have been resolved, the user is finally logged in.
 
 [[_auth_spi_walkthrough]]
 === Authenticator SPI Walk Through
 
 In this section, we'll take a look at the Authenticator interface.
-For this, we are going to implement an authenticator that requires that a user enter in the answer to a secret question like "What is your mother's maiden name?".  This example is fully implemented and contained in the examples/providers/authenticator directory of the demo distribution of {project_name}. 
+For this, we are going to implement an authenticator that requires that a user enter in the answer to a secret question like "What is your mother's maiden name?".
+This example is fully implemented and contained in the examples/providers/authenticator directory of the demo distribution of {project_name}.
+
+To create an authenticator, you must at minimum implement the org.keycloak.authentication.AuthenticatorFactory and Authenticator interfaces.
+The Authenticator interface defines the logic. The AuthenticatorFactory is responsible for creating instances of an Authenticator.
+They both extend a more generic Provider and ProviderFactory set of interfaces that other {project_name} components like User Federation do.
+
+Some authenticators, like the CookieAuthenticator don't rely on a Credential that the user has or knows to authenticate the user. But some, like the
```

Comment:
Change 
"But some, like the
PasswordForm authenticator or the OTPFormAuthenticator"
TO
"However, some authenticators, such as the
PasswordForm authenticator or the OTPFormAuthenticator, "